File: python/sparklabs/batch/sales/dataframe/revenue_per_store.py
from pyspark.sql import functions as F

from sparklabs.batch.sales.dataframe import utils


def run() -> None:
    revenue_per_order = (
        utils.order_items.join(utils.products, ["product_id"], "inner")
        .withColumn("revenue", F.col("price") * F.col("quantity"))
        .groupBy(F.col("order_id"))
        .agg(
            F.sum("revenue").alias("total_revenue"),
            F.count("*").alias("total_orders"),
        )
        .orderBy(F.desc("total_revenue"))
    )

    (
        utils.orders.join(revenue_per_order, ["order_id"], "inner")
        .join(utils.stores, ["store_id"], "inner")
        .groupBy(F.col("store_id"))
        .agg(
            F.sum("total_revenue").alias("store_revenue"),
            F.count("*").alias("total_store_orders"),
        )
        .orderBy(F.desc("store_revenue"))
        .show(truncate=False)
    )


if __name__ == "__main__":
    run()

File: r/sparklabs/batch/sales/dataframe/revenue_per_store.R
source("r/sparklabs/batch/sales/dataframe/utils.R")

revenue_df <- withColumn(
  join(order_items, products, order_items$product_id == products$product_id),
  "revenue",
  products$price * order_items$quantity
)
revenue_per_order <- agg(
  groupBy(revenue_df, revenue_df$order_id),
  total_revenue = sum(revenue_df$revenue),
  total_orders = count(revenue_df$order_id)
)
joined <- join(orders, revenue_per_order, orders$order_id == revenue_per_order$order_id)
joined <- join(joined, stores, joined$store_id == stores$store_id)
result <- orderBy(
  agg(groupBy(joined, joined$store_id), store_revenue = sum(joined$total_revenue), total_store_orders = count(joined$store_id)),
  desc("store_revenue")
)
showDF(result, truncate = FALSE)
